airflow/plugins/operators/sentry_to_gcs.py

```diff
@@ -0,0 +1,164 @@
+    to the type of data in the array.
+    """
+    types = set(type(entry) for entry in arr if entry is not None)
+
+    if not types:
+        return []
+    # use empty string for all non-numeric types
+    # may need to expand this over time
+    filler = -1 if types <= {int, float} else ""
+    return [x if x is not None else filler for x in arr]
+
+
+def make_arrays_bq_safe(raw_data):
+    safe_data = {}
+    for k, v in raw_data.items():
+        if isinstance(v, dict):
+            v = make_arrays_bq_safe(v)
+        elif isinstance(v, list):
+            v = process_arrays_for_nulls(v)
+        safe_data[k] = v
+    return safe_data
+
+
+def get_issues_list_from_sentry(extract, headers):
+    """
+    Paginate over Sentry issues for a project and create a list of issues matching our
+    criteria for examination (RTFetchExceptions).
+    """
+    next_url = BASE_URL + f"projects/sentry/{extract.project_slug}/issues/"
+    response_data = []
+
+    while next_url:
+        response = requests.get(next_url, headers=headers)
+        response_data.extend(response.json())
+        if response.links["next"]["results"] == "true":
+            next_url = response.links["next"]["url"]
+        else:
+            next_url = None
+
+    issues_list = [x["id"] for x in response_data if "RTFetchException" in x["title"]]
+    return issues_list
+
+
+def iterate_over_sentry_records(extract, auth_token):
+    """
+    Paginate over API responses for each targeted issue and create a combined list of dicts.
+    """
+    headers = {"Authorization": "Bearer " + auth_token}
+    issues_list = get_issues_list_from_sentry(extract, headers)
+    response_data = []
+
+    for issue_id in issues_list:
+        next_url = BASE_URL + f"issues/{issue_id}/events/"
+
+        while next_url:
+            response = requests.get(next_url, headers=headers)
+            response_data.extend(response.json())
+            if response.links["next"]["results"] == "true":
+                next_url = response.links["next"]["url"]
+            else:
+                next_url = None
+
+    return response_data
+
+
+def fetch_and_clean_from_sentry(extract, auth_token):
+    """
+    Download Sentry event records as a DataFrame.
+    """
+
+    print(f"Downloading Sentry event data for project {extract.project_slug}")
+    all_rows = iterate_over_sentry_records(extract, auth_token)
+    extract.extract_ts = pendulum.now()
+
+    raw_df = pd.DataFrame([{**make_arrays_bq_safe(row)} for row in all_rows])
+
+    cleaned_df = raw_df.rename(make_name_bq_safe, axis="columns")
+
+    extract.data = cleaned_df.to_json(orient="records", lines=True).encode()
+
+    return extract
+
+
+class SentryExtract(PartitionedGCSArtifact):
+    bucket: ClassVar[str] = CALITP_BUCKET__SENTRY_LOGS
+    table: ClassVar[str] = "events"
+    dt: pendulum.Date
+    ts: pendulum.DateTime
+    partition_names: ClassVar[List[str]] = ["dt", "ts"]
```

Comment:
I might recommend adding `project_slug` as a partition if you aren't adding it to the individual records; it looks like the records just contain the numeric id in `projectid` right now

Reply:
I think this is still relevant; I don't see anything in individual records other than the numeric `projectid`.

Actually taking this with partitioning by _event_ date, I think the best set of partitions might be `event_dt, project_slug` so data is clearly labeled by project in GCS as well. `extract_ts` could either be the final partition or the filename, just some way to distinguish between multiple extracts of the same data.